### ISSUE TITLE
Access to the log-density after training

### ISSUE THREAD

**Issue by philastrophist** [bgroenks96/normalizing-flows#1]
Hi @bgroenks96 

I'm fairly new to working with NNs and I've never used normalising flows before - hopefully you point me in the right direction. 
I'm trying to learn a non-analytical ~10 dimensional function (hopefully) and I want to be able have access to the logp of the flow at the end. This requires access to both forward and inverse transform as well as the jacobian.  I understand that I should be using Glow for this.

So really I have two questions:

1. How do I train a network on a function, not a dataset? So I'm not doing inference per se, just learning a function
2. How do I access logp after the training (not the loss, but the logp of the model)? 

Any hints would be greatly appreciated!
Thanks

**Comment by bgroenks96** [comment]
Hi @philastrophist 

1. Is your function spatial (e.g. images)? If not, then Glow probably isn't the right architecture to use. You could consider instead a variational autoencoder with the `TriangularSylvester` or `Planar` flows. There are of course lots of primitive operations that you can chain together to build a bijector (i.e. flow). You can find a lot more of them implemented in [Tensorflow Probability](https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors). I would also start with a more well supported package like TFP or pymc3 until you run into issues or limitations. Then you could consider building on this code instead.

2. Training directly on a function is not supported. But you might be able to accomplish this using the generator constructor of `tf.Dataset`, if you can find a way to build your function into the generator. You could also "pre-generate" a dataset from your function, if this is computationally feasible.

3. The `FLVM` and `JFLVM` models in this package both return the negative log likelihood which is probably what you're after. You can look there for examples. In general, all flow transformations return the output as well as the `logdetJ`, so you can always compute the transformed `logp` from the prior and this accumulated `logdetJ` term.

Hope that helps!

**Comment by philastrophist** [comment]
Thanks that's very helpful!
